scorex-basics/src/main/scala/scorex/network/peer/PeerDatabaseImpl.scala

```diff
@@ -21,18 +22,30 @@ class PeerDatabaseImpl(settings: Settings, filename: Option[String]) extends Pee
   private lazy val ownNonce = settings.nodeNonce
 
   override def addOrUpdateKnownPeer(address: InetSocketAddress, peerInfo: PeerInfo): Unit = {
-    val updatedPeerInfo = Option(whitelistPersistence.get(address)).map { case dbPeerInfo =>
-      val nonceOpt = peerInfo.nonce.orElse(dbPeerInfo.nonce)
-      val nodeNameOpt = peerInfo.nodeName.orElse(dbPeerInfo.nodeName)
-      PeerInfo(peerInfo.lastSeen, nonceOpt, nodeNameOpt)
-    }.getOrElse(peerInfo)
-    whitelistPersistence.put(address, updatedPeerInfo)
-    database.commit()
+    if (!isBlacklisted(address)) {
+      val updatedPeerInfo = Option(whitelistPersistence.get(address)).map { case dbPeerInfo =>
+        val nonceOpt = peerInfo.nonce.orElse(dbPeerInfo.nonce)
+        val nodeNameOpt = peerInfo.nodeName.orElse(dbPeerInfo.nodeName)
+        PeerInfo(peerInfo.lastSeen, nonceOpt, nodeNameOpt)
+      }.getOrElse(peerInfo)
+      whitelistPersistence.put(address, updatedPeerInfo)
+      database.commit()
+    }
   }
 
   override def blacklistPeer(address: InetSocketAddress): Unit = {
-    whitelistPersistence.remove(address)
-    if (!isBlacklisted(address)) blacklist += address.getHostName -> System.currentTimeMillis()
+    if (!isBlacklisted(address)) {
+      blacklist += address.getHostName -> System.currentTimeMillis()
+      new util.ArrayList(whitelistPersistence.keyList())
```

Comment:
I started from this version, and it did not work - some of removed elements in fact retained in the whitelistPersistence which causes test fails etc. The possible reason is modification of the collection we iterate through. From the Java Iterator.remove() docs: "The behaviour of an iterator is unspecified if the underlying collection is modified while the iteration is in progress in any way other than by calling this method". Since we have here a Java collection, I have decided to use a quick Java-style approach - just to create a copy of the collection.